### ISSUE TITLE
`useContext` and `createContext` might be handy

### ISSUE THREAD

**Issue by WebReflection** [WebReflection/augmentor#2]
Coming from this conversation https://github.com/WebReflection/neverland/issues/6

It might actually make sense to have such mechanism in here too.

**Comment by dakom** [comment]
Btw doing more homework into the actual goal I'm thinking about... moving from React+whatever to hyper+xstate - this looks really elegant at a glance: https://github.com/carloslfu/use-machine

Seems like it can be done with neverland/augmentor once you add useContext/createContext

Without context, it's pretty much a no-go because it makes communicating between the statecharts too complicated

**Comment by dakom** [comment]
FYI - I asked around and evidently the idea of context _is_ that it's an optimization over just passing props down in that only the components which use the context get re-rendered when the context changes (not the intermediate parents)

**Comment by WebReflection** [comment]
I've sneaked in a version with `useContext`.

Since I'm not super familiar with it, I'd like to give It a spin before making it official in the README / documentation.

Can you please see if it works as expected ?

It does [here](https://webreflection.github.io/augmentor/test/context.html), as example, [from this file](https://github.com/WebReflection/augmentor/blob/master/test/context.html).

Neverland uses this latest version with `createContext` and `useContext` too.

**Comment by joshgillies** [comment]
Amazing work! Will take a look over a few examples today to see how it fairs!

**Comment by dakom** [comment]
FYI Part 2... the following is wrong:

> FYI - I asked around and evidently the idea of context is that it's an optimization over just passing props down in that only the components which use the context get re-rendered when the context changes (not the intermediate parents)

Rather, context _does_ force a re-render down the tree because there's no way of passing a changed value down to the consumer without changing it at the top-level state, so it effectively forces a re-render.

To get around it and stop intermediate parents from rendering, you have to use other React primitives (shouldComponentUpdate in the class system, memo() in the new hooks version)

So to achieve the same thing, we need something like `memo()`

Here's a React version I made which demonstrates the bug (items without a Tick should _not_ re-render): https://codepen.io/dakom/pen/REOzxp

And here's a fork someone made which fixes it via `React.memo()`: https://codepen.io/eligolding/pen/ebaVNv

Once `useContext/createContext` are [exported for Neverland](https://github.com/WebReflection/neverland/issues/6#issuecomment-455234026) I'm happy to do a port  to demo it (also an easy example to get used to the system and hone my chops!) - but someone else gets to it first to knock it out quickly, by all means :)

**Comment by WebReflection** [comment]
I won't have time to do much today, but these demos are very valuable. Thanks for writing and linking 👍

**Comment by dakom** [comment]
I think you've got it... I've like 99% converted the demo over here:

https://codepen.io/dakom/pen/QzRmar?editors=0010

Only problem is creating a proper `memo()` function... I'm really not sure if `useMemo` the way I did it in the commented out part is even correct, though it is odd that it bugs out... I'll open a separate ticket for that.

What is cool is that by freezing the top-level renders via useMemo and an empty array (as it is now), and passing context down - we can see that it's rendering the children via context even when the parents aren't re-rendering.

So in short, LGTM! 

Well done and thanks @WebReflection !!

**Comment by WebReflection** [comment]
> So in short, LGTM!

Something ain't quite right though, 'cause the whole dom is re-rendered each time, while I'd expect only the tick numbers to be modified per each update.

Some reference gets lost in the process, will try to figure out why/how.

Btw, is the `useMemo` broken? In that case I'd like to fix that before the experimental `useContext`.

**Comment by dakom** [comment]
Yeah - I see it's also just updating the text... interesting.

Re: permalink to unpkg, sure.. maybe it's worth having a list of permanent-ish integration tests like this on one of the repo wiki's?

I'm happy to go in and update the dependencies. Gotta run now though, will followup later!
